result_collector.go

```diff
@@ -0,0 +1,137 @@
+package wire
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+)
+
+// Pipeline safety limits
+const (
+	MaxPipelineQueries = 20
+	MaxPipelineRows    = 50000
+)
+
+// ResultCollector implements DataWriter interface
+// It collects query results in memory for later replay
+type ResultCollector struct {
+	mu      sync.Mutex
+	columns Columns
+	rows    [][]any
+	tag     string
+	empty   bool
+	written uint32
+	err     error
+	limit   Limit
+}
+
+// Implement DataWriter interface
+
+func (rc *ResultCollector) Row(values []any) error {
+	rc.mu.Lock()
+	defer rc.mu.Unlock()
+
+	if rc.err != nil {
+		return rc.err
+	}
+
+	rc.rows = append(rc.rows, values)
+	rc.written++
+
+	// Check if we've exceeded the row limit for pipeline mode
+	if len(rc.rows) >= MaxPipelineRows {
+		rc.err = fmt.Errorf("query result too large for pipeline mode (%d rows exceeds limit of %d)", len(rc.rows), MaxPipelineRows)
+		return rc.err
+	}
+
+	return nil
+}
+
+func (rc *ResultCollector) Complete(tag string) error {
+	rc.mu.Lock()
+	defer rc.mu.Unlock()
+
+	rc.tag = tag
+	return nil
+}
+
+func (rc *ResultCollector) Empty() error {
+	rc.mu.Lock()
+	defer rc.mu.Unlock()
+
+	rc.empty = true
+	return nil
+}
+
+func (rc *ResultCollector) Columns() Columns {
+	rc.mu.Lock()
+	defer rc.mu.Unlock()
+
+	return rc.columns
+}
+
+func (rc *ResultCollector) Written() uint32 {
+	rc.mu.Lock()
+	defer rc.mu.Unlock()
+
+	return rc.written
+}
+
+func (rc *ResultCollector) CopyIn(format FormatCode) (*CopyReader, error) {
+	return nil, errors.New("CopyIn not supported in pipeline mode")
+}
+
+func (rc *ResultCollector) Limit() uint32 {
+	return uint32(rc.limit)
+}
+
+// SetError sets the error state (thread-safe)
+func (rc *ResultCollector) SetError(err error) {
+	rc.mu.Lock()
+	defer rc.mu.Unlock()
+	rc.err = err
+}
+
+// GetError gets the error state (thread-safe)
+func (rc *ResultCollector) GetError() error {
+	rc.mu.Lock()
+	defer rc.mu.Unlock()
+	return rc.err
+}
+
+// Replay writes all collected data to a real DataWriter
+func (rc *ResultCollector) Replay(ctx context.Context, writer DataWriter) error {
+	rc.mu.Lock()
+	defer rc.mu.Unlock()
+
+	if rc.err != nil {
+		return rc.err
+	}
+
+	if rc.empty {
+		return writer.Empty()
+	}
```

Comment:
Do we have to keep track of the empty state using the `empty` property? The empty method is right now called only if no columns have been written.

```go
	if writer.written == 0 && writer.columns != nil {
		err := writer.Empty()
		if err != nil {
			return err
		}
	}
```